### PR TITLE
Deprecate jacobian tape

### PR DESCRIPTION
**Pull request by josh146**
**Context:** The tape subclasses and `tape.execute()` are no longer used by PennyLane, and ready for a deprecation cycle.

**Description of the Change:** All tape subclasses and execution methods have been marked as deprecated.

**Benefits:** n/a

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/gradients/gradient_transform.py [hunk 1]
```diff
@@ -20,6 +20,60 @@
 from pennylane.transforms.tape_expand import expand_invalid_trainable
 
 
+def gradient_analysis(tape, use_graph=True, grad_fn=None):
```
Comment (josh146) [inline pennylane/gradients/gradient_transform.py:23]: This function was moved from `pennylane.gradients.parameter_shift._gradient_analysis`, since it turns out that `finite_diff` also needed to use it.

Comment (antalszava) [reply pennylane/gradients/gradient_transform.py:23]: Out of scope: the name `gradient_analysis` seems to suggest that we're getting information about the gradient of the tape.

At the same time:
> Note that this function modifies the input tape in-place.

I think seeing a line like
```python
gradient_analysis(tape, grad_fn=finite_diff)
```
and mutating the tape data under the hood is unexpected. Any reasons why the `_update_gradient_info` name was abandoned? It contained a verb that referred to the updating.

File: pennylane/gradients/vjp.py [hunk 2]
```diff
@@ -145,18 +144,18 @@ def vjp(tape, dy, gradient_fn, gradient_kwargs=None):
...
+             2.1329e-17], dtype=torch.float64, grad_fn=<ViewBackward>)
 
     The output VJP is also differentiable with respect to the tape parameters:
 
     >>> cost = torch.sum(vjp)
     >>> cost.backward()
     >>> x.grad
     tensor([[-1.1025e+00, -2.0554e-01, -1.4917e-01],
-            [-1.9429e-09, -9.1580e-01,  1.3878e-09]], dtype=torch.float64)
+            [-1.2490e-16, -9.1580e-01,  0.0000e+00]], dtype=torch.float64)
```
Comment (antalszava) [inline pennylane/gradients/vjp.py:158]: Could this have been bugs at the time or something changed internally? Seems like params were fixed.

File: pennylane/tape/tape.py [hunk 3]
```diff
@@ -254,19 +255,10 @@ class QuantumTape(AnnotatedQueue):
...
-    Execution can take place either using the in-place constructed parameters,
-
-    >>> tape.execute(dev)
-    [0.77750694]
-
-    or by providing parameters at run time:
-
-    >>> tape.execute(dev, params=[0.1, 0.1, 0.1])
-    [0.99003329]
+    >>> qml.execute([tape], dev, gradient_fn=None)
```
Comment (antalszava) [inline pennylane/tape/tape.py:261]:
```suggestion
    >>> qml.execute([tape], dev, gradient_fn=None)
    [array([0.77750694])]
```

File: tests/gradients/test_parameter_shift.py [hunk 4]
```diff
@@ -28,52 +28,53 @@ def test_non_differentiable(self):
...
+        with qml.tape.QuantumTape() as tape:
             qml.QubitStateVector(psi, wires=[0, 1])
             qml.RX(0.543, wires=[0])
             qml.RY(-0.654, wires=[1])
             qml.CNOT(wires=[0, 1])
             qml.probs(wires=[0, 1])
 
         spy = mocker.spy(qml.operation, "has_grad_method")
-        _gradient_analysis(tape)
+        gradient_analysis(tape, grad_fn=5)
```
Comment (antalszava) [inline tests/gradients/test_parameter_shift.py:57]: Why pass `grad_fn=5`?

**Review by antalszava — APPROVED** [review]
Looks good :+1: :tada: Just had some minor questions/suggestions.

File: pennylane/gradients/gradient_transform.py (hunk 1 above)
Comment (josh146) [reply pennylane/gradients/gradient_transform.py:23]: @antalszava yep 😆 Agree on all points. `gradient_analysis` is actually pretty much a 1-1 port of `tape._update_gradient_info()`, hence why it is in-place - it was originally a Tape method!

In terms of why it was renamed, I can't remember. Happy for it's name to be changed, but would rather it be done in a separate PR

Comment (josh146) [reply pennylane/gradients/gradient_transform.py:23]: Although rather than renaming it, we should probably refactor this function to (a) not be in place, and (b) have a smarter/cleaner form of caching rather than adding a new attribute to the tape.

File: pennylane/gradients/vjp.py (hunk 2 above)
Comment (josh146) [reply pennylane/gradients/vjp.py:158]: This is a good question 🤔 Let me double check.

Comment (josh146) [reply pennylane/gradients/vjp.py:158]: @antalszava looks like this was a bug at the time :) I re-ran the original code examples, and get the same output as above.

File: pennylane/tape/tape.py (hunk 3 above)
Comment (josh146) [reply pennylane/tape/tape.py:261]: oops! thanks for catching this.

File: tests/gradients/test_parameter_shift.py (hunk 4 above)
Comment (josh146) [reply tests/gradients/test_parameter_shift.py:57]: `gradient_analysis` has behaviour that caches the analysis iff:

- the tape is the same
- the `grad_fn` is the same.

This way, you can run `gradient_analysis(tape, param_shift)` multiple times while avoiding overhead. But if you _change_ the gradient function, then the analysis will be re-run.

It is a bit ugly, and the design of `gradient_analysis` should be re-thought, but I thought out of scope for this PR.

> Why pass grad_fn=5?

I thought it would be better test design then doing `grad_fn=qml.gradients.param_shift`, which could be an annoying point of test failure if we move/change this function in the future!
